### DISCUSSION TITLE
Native: Globe Mode Proposal

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#320] [Native feature proposals]
### Globe View with Web Mercator Sources For Native

Here's some functionality we'd like to add to MapLibre.  This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals).  Short version: Native first because we know it, then circle back and propose web.

We're talking about a simple globe view with web mercator data sources.  If this is interesting, let's discuss.  If you have questions, please ask.

## Top Down Globe View

This a (mostly) top down globe view.  It would be sufficient to display worldwide, or large data sets, imagery, weather, and that sort of thing.  That would look something like this.

![Nat Geo World Atlas](https://github.com/maplibre/maplibre/assets/101820560/b3477911-d089-4d22-af2a-bcbd4dd1e185)

It would expect one data set as a base, such as a map tile pyramid or imagery data set that covers the whole globe, with the possible exception of the poles.  Additional data sets could be layered on top.  Here's an example.

The poles would be handled separately, by default pulling colors from the nearest northern and southernmost tiles.  Here's what the that looks like.

![North pole example](https://github.com/maplibre/maplibre/assets/101820560/f33583e6-abcd-444e-acca-5d196d3bf1b3)

## Gestures and Tilt
Gesture support would be implemented with a standard Quaternion library.  They say it's best to implement your own Quaternion, then throw it away and use someone else's.  We've already done that, saving time on this project.

Interaction would feel like the standard floating bowling ball and support free rotation, North Up, and Track Up.  This can get very fiddly, so we'd take a round of user feedback and add more to it.  Ideally we'd open these up for developer fiddling as well.

The only 'grabbable' object will be the globe and all interaction will be based around it.  Tilt will work up to a certain point, beyond which it will just be annoying.

## Data Sources

The project would work with existing data sources such as image tiles, vector tiles, and GeoJSON.  Only data in Web Mercator (for tiles) and lat/lon (for GeoJSON) will work initially, but we have another proposal for that.

Image tiles will be plastered on top of the globe, or used as the base image.  Vectors will be draped on top and may be chopped up to follow the globe's curvature.

Vector tiles present a challenge as they don't work well at low zoom levels.  Things come out very... polygonal.

The solution we'd use is to render the polygons into an image per tile and drape the linear vector on top.  It's a rather fiddly solution that does best with a bit of style sheet support to decide what goes where.  It can be made to work.

![Carrot Weather Globe](https://github.com/maplibre/maplibre/assets/101820560/8257b4a8-9e8a-410b-a101-b05589e77bae)

## Data Loading

One of the harder problems on the globe is deciding what to load.  We have experience doing it with arbitrary coordinate systems on the fly and let me tell you... it's a lot.  So we're starting simple here.

Web Mercator is easily converted to and from geocentric (fancy way of saying the globe).  So we restrict to that for the moment.  We'll probably do something hacky and simple to figure out what to load initially.  Later on... well we have a proposal on that.

## Discussion

The obvious question is why not full 3D?  Why not support any data projection from anywhere?  Why not do it all at once?

The answer to that last question is always risk.  We want to give you the option of deciding what features are important or not rather than doing everything at once.

As to the first, full 3D and globe 3D are actually two different things.  The interactions between viewer and data are very different when spinning a globe versus moving through a building.  Look closely at a viewer that does both and you'll realize it's quietly switching modes on the fly.

Lastly, why only Web Mercator?  Well, it's simple.  Doing any map projection in and out is a different problem.  Being precise enough for something like aviation... is a different problem.  So we start simple.

**Comment by HarelM** [comment]
A few things to take into consideration:
1. Labels - these should not be part of the vector imaging drape - they should be kept straight/according to the style, collision etc should work as it works now.
2. "image source" - we have a problem with it in web where the image gets cut on tiles edge when terrain is on, I would recommend solving this by creating a tile pyramid from the image and then use it like a regular raster source, I don't know how it is implemented in native right now, so this is just in general.
3. Adaptive projection from globe to the current 2D map is what I would expect as a default experience
4. Stars in the background? might be interesting as a good experience, totally nice to have.
5. Atmosphere - I've played with cesium and it has a lot of parameters there that are creating cool stuff, there's also a PR in the spec about sky and fog implementation, not sure if and how it is related to this issue, but worth taking it into consideration.

Thanks for opening this discussion!

**Reply by sjg-wdw** [reply]
For labels, right, you want to drape the points and then generate normal labels (we called them screen labels) from there.  The layout logic gets a bit more complex, no doubt.

Ah right, single images will be tricky.  In WhirlyGlobe we had the concept of "stickers" which were similar.  We can borrow some of the logic from that.

For adaptive projection, do you mean animation to and from a 2D map to a 3D globe?
That's actually fairly tricky to do well.

Stars and Atmosphere - I think a separate proposal might be good for that one.  I'll make a note.

**Reply by HarelM** [reply]
I mean that you see a globe and as you zoom in the projection changes from a globe projection to mercator projection.
Something like the following animation, but obviously better:
https://github.com/maplibre/maplibre/discussions/161

**Reply by birkskyum** [reply]
**Adaptive Globe Projection** generally means 3D globe at low zoom -> web mercator on high zoom level, similar to the default view of Apple Maps and Google Maps (web version). [This is our most requested feature](https://github.com/maplibre/maplibre/discussions/categories/bounties-cross-project-directions?discussions_q=is%3Aopen+category%3A%22Bounties+-+cross-project+directions%22+sort%3Atop), and it has potential to become a new default mode.

**Adaptive 2D Projections**, such as animating from Natural Earth projection on low zoom to web mercator on high zoom is much less requested compared to the adaptive globe, but if the work around "zoom to mercator" from the adaptive globe can make it more feasible to add, it might be worth considering down the line.

**Reply by HarelM** [reply]
Oh, and it's worth mentioning that dynamic hill shading with all that 3D (globe or not) is a nightmare...

**Reply by sjg-wdw** [reply]
I figured that's what you meant.  It's not easy to do well, particularly without controlling the visual database.

But heck, that's the point of asking.  See what people want.  Let's see what we can do!

**Reply by birkskyum** [reply]
@HarelM , @sjg-wdw . Believe I may have said something wrong here, so i'd like to correct myself - Despite what I thought was the case, it comes out that google maps and apple maps in globe view actually doesn't adaptively adjust to mercator on high zoom, so it's not an industry standard in any way at this point. In globe view, apple/google are just globes, and because of the size of Earth the distortion on high zoom is relatively small, given that it's somewhat hard to even verify that it's there. I checked by going to high zoom in Google Maps and toggled the globe view - or if lat/lon lines can be shown, it can simply be checked if they are straight (i.e. google earth).

Another aspect is that for GL JS, I think starting with a simple globe would be much easier to make compatible with ecosystem efforts like i.e. [Deck.GL experimental globe view](https://deck.gl/docs/api-reference/core/globe-view), because interleaved data layers would have to mirror all projection changes in sync with maplibre. @Pessimistress , just fyi that globe compat. is being thought about on the maplibre side.

**Reply by sjg-wdw** [reply]
I kind of thought that's how those maps worked, but I hadn't looked very closely recently.  That sort of morphing functionality is cool, but I worry that it add burdens to new layer developers.  So if it's done, I'd want to see it done so they don't have to implement it.

Compatibility with Deck.GL is an interesting problem.  Worth a look before starting, if this effort moves forward.  I'm very sympathetic to this kind of thing.  We do an overlay of our own rendering toolkit on top of MapLibre JS for the web and appreciate that the matrix and render time callbacks are available.

**Comment by birkskyum** [comment]
@sjg-wdw , I know we have to keep it for everything but iOS/macOS, but just to understand, is there going to be a significant cost to keeping support for the OpenGL target when adding new features like this, compared to only having to deal with more modern graphics APIs like Metal, or does that not make a difference with the introduction of the drawables render and the bump from OpenGL 2 to a newer version ?

**Reply by sjg-wdw** [reply]
Got to keep the OpenGL renderer.  That's what's used on Linux for offline rendering, a pretty important use case for at least one big consortium member.

It's now easier to drop in a new renderer.  We just did it, of course, so that's reassuring.  From what I saw before the project was having trouble doing much to the renderer at all, so I'd say the situation has improved.  The problem before was sheer complexity.  It wasn't meant to be user serviceable, as it were, and now it is.

So each new rendering SDK will have a cost in maintenance.  That's very true.  It would be cheaper to get rid of an SDK or not add a new one.  Or.... you could just lean into it.  Keep that interface to rendering SDKs up to date and you can survive the next transition, whatever that may be.

**Reply by birkskyum** [reply]
I'd love to know from this "one big consortium member" if Vulkan support on the hardware is widely in place, or if it's still a pain point.

**Reply by maxammann** [reply]
We want to keep some OpenGL support probably indefinitely. E.g. the Android Emulator requires any application to support egl/GL. Also there are still phones out there that do not support Vulkan.

**Reply by sjg-wdw** [reply]
That particular use case is a software renderer.  Would Vulkan work for that?  Maybe, but it's effort for the user.

So then you're asking a big user to upgrade to something that doesn't give them any benefit.  Which means they won't bother to upgrade.  Now they're on an old version and can't see any reason to upgrade to the new version.  Not where you want them, particularly if they have resources for you.

Dropping functionality is always a valid conversation to have, but the cost is much higher than it seems.  Even GDAL, who has massive power over its base, got a lot of pushback for deleting support for a weird IBM database no one uses.  Feelings were hurt.  In their case, they had a really good reason, but it was still painful.

My recommendation is to resign yourself to supporting OpenGL ES for another decade.  With the new Native rendering architecture I don't think it even costs you that much.

**Comment by wipfli** [comment]
Can you give an estimate on binary size impact @sjg-wdw?

**Reply by sjg-wdw** [reply]
Honestly, no.  Don't know how to do that.
It would certainly be bigger if it went into the main library rather than an auxiliary.

**Reply by wipfli** [reply]
Do you think there is a way to make the features you are proposing opt-in?

Because there are some users of MapLibre Native who will have a very hard time justifying any size increase to their leadership. From what I understood, larger app means less users will download it from the app store...

**Reply by sjg-wdw** [reply]
Oh yes, I think we want to make almost everything opt-in.  Too many important users won't adopt a bigger toolkit.

**Reply by wipfli** [reply]
I am glad to hear that, thanks!

**Comment by wipfli** [comment]
Not sure where the best place is to ask questions on all 4 proposals, so let me just do it here:

- Do you think it is better to change MapLibre Native and add new things to it, or would it make more sense to re-write the toolkit from scratch?
- Do you think that if we were to rewrite the toolkit from scratch, there would be any chance that we could deploy it on iOS, Android, and web browsers?

**Reply by birkskyum** [reply]
There are ways we can reach web browsers even without a rewrite. 
- #339 
- https://github.com/maplibre/maplibre-native/issues/454

**Reply by sjg-wdw** [reply]
Ooo, what an interesting question.

From a technical perspective, perhaps it would be better to start over from scratch.  Maybe use WebAssembly for some key pieces and certainly design from a more modern perspective.

From a business perspective, disaster.  You don't have the budget to both support (and update!) the existing toolkits while rewriting them.

From a user perspective, also disaster.  You're inheriting an existing community.  If you want to continue inheriting that community, I'd suggest you be calm and soothing and offer easy upgrade paths.  Not tear it all up.  You'd have to talk that community into something new and they're not really in that mood.

From an engineering management perspective, it's courting disaster.  Rewriting everything from scratch sounds good, but there are so many ways to fail and so many, many hilarious tales of failure.  The more mature organizations are going to recognize this and start hedging their risk.

With this kind of thing I like to step back and ask where the urge to rewrite is coming from.  Maybe it's the urge to do something new, to try out new technologies and add new functionality.  To fill a gap in the offerings.  Well, there are other ways to do that.

You've got a squishy, organic Javascript version with some popularity.  Don't try to turn that into a robot (cross compiled WASM).  Maybe turn it into a cyborg (some functionality in WASM).  

Try adding a new piece on the native side, cross compile it and insert it dynamically on the web side.  How does that work out?  If the community rejects it, well okay then rewrite it in Javascript.  If the community likes it, then maybe that's a path forward.  Javascript is more friendly to these sorts of additions, so do it in a way JS devs find comfortable.

In general, incrementalism is better, particularly when you don't have a massive budget.  That's my opinion.

**Reply by sjg-wdw** [reply]
You do have a big risk on the lower/faster/simpler end of things.  I've got another proposal to talk about that.

**Reply by HarelM** [reply]
Not really related to the title, but in continue to this thread the following pieces might be an interesting POC for code sharing between native, web (and maybe maplibre-rs):
1. Geojson to vector tiles conversion, currently done in a worker in maplibre-gl-js and uses geojson-vt package, I think it would be an interesting thing to try out
2. Style spec validation, maybe should be part of the style-spec repo
3. Labels collision - although this might require too much rendering logic, so I would prioritize this as last

CC: @louwers @maxammann

**Reply by louwers** [reply]
Code sharing needs to bring some benefit for both Native and GL JS, because the build tooling overhead and added complexity is likely to be pretty significant. Benefits I can think of are:

1. making it easier to keep functionality in sync
2. not having to implement the same thing twice
3. increased performance

Thus to find a good use case for code sharing I would rather identify something that:

1. Is likely to receive changes and improvements over time.
2. Is new functionality (not already implemented).
3. Benefits from the added performance of native code.

Of course we can try out code sharing with something like geojson-vt, but I doubt it will make it into production if the benefits are not clear.

**Reply by HarelM** [reply]
What I'm interested to know is how complicated it is and if it's even feasible or not. I don't mind not having it in production as long as we understand what parts are relevant for code sharing if there are any.
I think geojson vt is a great way to POC this.

**Reply by sjg-wdw** [reply]
I think @louwers' checklist is spot on.  

I don't know what the fate of our various proposals will be, but I think we could tack on a requirement to cross-compile to WASM for some of them.  Particularly the ones dealing with a globe and coordinate systems.  

We might have mentioned this, but we did cross-compile WhirlyGlobe to WASM and use it in our weather displays for the web.  We're comfortable enough with Emscripten to use it again.

**Reply by maxammann** [reply]
I'm a little surprised that maplibre-rs was only mentioned along the side here, as this is the core idea of the MapLibre Rust experiment: Cross-platform rendering using WebGPU and WebAssembly.

We could drop support for the OpenGL renderer by directl supporting WebGPU instead of Metal. We can then use the Dawn engine to translate from WebGPU to Metal with almost zero-cost. This could be one deprecation strategy for the OpenGL backend.

---

I'm sceptical about translating parts of maplibre to Rust/C++ and reuse them in native or web as WebAssembly. There is currently not that much modular code that can be reused. The geojson-vt could be interseting, though it is also not really a problem that it is not reused. It is hard to come up with good targets for code-sharing imo.

---

> From a business perspective, disaster. You don't have the budget to both support (and update!) the existing toolkits while rewriting them.

I agree, it is quite an investment (>1000000$).

> From a user perspective, also disaster. 

If everything would go right, users would not notice except for some cases like DeckGL support.

> From an engineering management perspective, it's courting disaster.

We already have 2 implementation of maplibre that work well. A rewrite would only be a translation of something that is known to work well. The maplibre-rs project already proved that WebGP, WebAssembly and Rust are tools that are a good fit for map rendering.


I definitely agree that it is a crazy investment, but MapLibre shoudl also consider if it is possible to bring new features at roughly the same time to the web and native, where both have very different technology stacks and even graphics APIs.

**Reply by HarelM** [reply]
This isn't the main topic of this discussion and thus the late mentioning of maplibe-rs.

>If everything would go right, users would not notice except for some cases like DeckGL support.

I don't think it will be an easy in place replacement, but I like your optimism.

> There is currently not that much modular code that can be reused.

This only means we are doing a bad job at trying to share code...
If we try and make maplibre micro service architecture like and move things around I'm sure we can share code.
Another example is the workers code, everything that is sent to the workers is serialized and serialized in the workers, if we can create this logic in rust and share it, it would pave the way forward I believe...
